### PR TITLE
[Feature] Density matrix partial trace

### PR DESCRIPTION
**Pull request by EthanObadia**
This pull request introduces **for later a draft version** of the `dm_partial_trace` function, which computes the partial trace of a density matrix for a system of several qubits with batch size.
The function also permutes the qubits according to the order specified in keep_indices.

### CODE REVIEW

**Comment by Roland-djee** [comment]
@EthanObadia What's the status here ?

**Comment by Roland-djee** [comment]
Right, OK. So @rajaiitp could you please review ? I think it makes sense to have it in PyQ because this is the kind on functionality the runtime in Qadence 2 will provide.

File: tests/test_digital.py [hunk 1]
```diff
@@ -382,3 +383,18 @@ def test_kron_batch() -> None:
...
+
+
+def test_dm_partial_trace() -> None:
+    n_qubits = torch.randint(low=1, high=5, size=(1,)).item()
+    batch_size = torch.randint(low=1, high=5, size=(1,)).item()
+    state_str = "".join(random.choice("01") for _ in range(n_qubits))
+    rho = density_mat(product_state(state_str, batch_size=batch_size))
+    keep_indices = random.sample(range(n_qubits), k=random.randint(1, n_qubits))
+    n_keep = len(keep_indices)
+    state_reduice_str = "".join([state_str[i] for i in keep_indices])
```
Comment (Roland-djee) [inline tests/test_digital.py:395]: Typo.

**Review by Roland-djee — APPROVED** [review]
Thanks @EthanObadia. LGTM.

**Comment by rajaiitp** [comment]
Sorry, this was totally outside my radar. Feel free to ping me if I fail to respond.

File: tests/test_digital.py (hunk 1 above)
Comment (EthanObadia) [reply tests/test_digital.py]: Changed.

File: tests/test_digital.py [hunk 2]
```diff
@@ -382,3 +383,18 @@ def test_kron_batch() -> None:
...
+    rho_reduice = dm_partial_trace(rho, keep_indices)
+    assert rho_reduice.shape == torch.Size([2**n_keep, 2**n_keep, batch_size])
+    assert torch.allclose(
+        rho_reduice, density_mat(product_state(state_reduice_str, batch_size=batch_size))
+    )
```
[the remark this answers is not in the document]
Comment (EthanObadia) [reply tests/test_digital.py:627]: Thank you for your remark. I will add a test to verify that the partial trace of the identity matrix is also an identity.

Regarding the index permutation, the current test already verifies this. Let's do an instance of this test:

For `n_qubits= 4`,  `state_str` can returns the string `0110` and `keep_indices` can returns the list `[2, 3, 1]` meaning that the `state_reduce_str` will be `101` as we keep in this order the 3rd, the 4th and the 1st qubit of the state `0110`. Finally, at the end of the test we assert that:

$$
\rho_{0110.reduce.permute} = \rho_{101} 
$$

[the remark this answers is not in the document]
Comment (Roland-djee) [reply tests/test_digital.py:633]: Would it also make sense to check the partial trace for the expectation of an operator acting on two sub-systems A and B ? Say:

$$
\langle O_A \rangle = tr_A(\rho_{AB}O)
$$

**Review by Roland-djee — COMMENTED** [review]
Hey @EthanObadia Looking good. Maybe expand a bit more on the tests ?

**Comment by EthanObadia** [comment]
I am no longer in charge of this pull request.  After syncing with @rajaiitp , he has kindly taken over the development. But, I can still provide my opinion.

File: tests/test_digital.py [hunk 3]
```diff
@@ -7,7 +7,9 @@
 import pytest
 import torch
 from conftest import _calc_mat_vec_wavefunction
-from torch import Tensor
+from numpy import array, rollaxis, sort
+from qutip import Qobj
```
[the remark this answers is not in the document]
Comment (rajaiitp) [reply tests/test_digital.py:11]: Tbh I'm not sure if I need to check this somewhere,
but the conftest.py  file in Pyqtorhc does import qutip.

**Comment by rajaiitp** [comment]
For some reason I cant respond to the above comment https://github.com/pasqal-io/pyqtorch/pull/186#discussion_r1672070160 directly, so addressing here.
We have verified using qutip for several instances using a batch for randomly generated density matrices and randomly selected subsystem qubits that the partial trace works fine, so I think it should be fine. We could increase the qubit system size if thats of interest, currently the upper limit is 5 qubits.

**Comment by Roland-djee** [comment]
> For some reason I cant respond to the above comment [#186 (comment)](https://github.com/pasqal-io/pyqtorch/pull/186#discussion_r1672070160) directly, so addressing here. 

Maybe because you have to un-draft and mark it as ready ? 

>We have verified using qutip for several instances using a batch for randomly generated density matrices and randomly selected subsystem qubits that the partial trace works fine, so I think it should be fine. We could increase the qubit system size if thats of interest, currently the upper limit is 5 qubits.

OK yes just push it a little. Thanks

**Review by rajaiitp — APPROVED** [review]

**Comment by rajaiitp** [comment]
Can I merge this ?

**Review by awennersteen — APPROVED** [review]
@rajaiitp 
please merge. If you don't have the rights just ping me on slack and I'll hit it

**Pull request merged**